File: ruby/lms.rb
#!/usr/bin/env ruby
require 'sinatra'
require 'webrick'
require 'json'

set :server, 'webrick'
set :port, 50139

def get_body(request)
  encoding = request.env['HTTP_CONTENT_ENCODING']
  request.body.rewind
  request_body = request.body.read
  # deflate gzip body
  request_body = Zlib::Inflate.new(window_bits=16+Zlib::MAX_WBITS).inflate(request_body).to_s if !encoding.nil? && encoding == 'gzip'
  JSON.parse request_body
end

post '/ruby/matching-service', :provides => 'application/json' do
  body = get_body(request)

  return JSON.generate({ result: "no-match" }) if body['levelOfAssurance'] != 'LEVEL_2'

  if !body['cycle3Dataset'].nil?
    return JSON.generate({ result: "no-match" }) if body['cycle3Dataset']['attributes']['nino'] == 'badvalue'
  end

  return JSON.generate({ result: "no-match" }) if body['matchingDataset']['surnames'][0]['value'] == 'Griffin'

  JSON.generate({ result: "match" })
end

post '/ruby/account-creation', :provides => 'application/json' do
  body = get_body(request)

  return JSON.generate({ result: "failure" }) if body['levelOfAssurance'] != 'LEVEL_2'

  JSON.generate({ result: "success" })
end

File: nodejs/lms.js

const http = require('http')

const hostname = '0.0.0.0'
const port = 4063

function localMatchingService (request) {
  if (request['levelOfAssurance'] !== 'LEVEL_2') {
    return { result: 'no-match' }
  }
  if (request['cycle3Dataset'] !== null &&
      request['cycle3Dataset'] !== undefined &&
      request['cycle3Dataset']['attributes'] !== null &&
      request['cycle3Dataset']['attributes']['nino'] === 'goodvalue') {
    return { result: 'match' }
  }
  if (request['matchingDataset'] !== null &&
      request['matchingDataset'] !== undefined &&
      request['matchingDataset']['surnames'] !== null &&
      request['matchingDataset']['surnames'][0]['value'] === 'Griffin') {
    return { result: 'match' }
  }
  return { result: 'no-match' }
}

function accountCreationService (request) {
  if (request['levelOfAssurance'] === 'LEVEL_2') {
    return { result: 'success' }
  }
  return { result: 'failure' }
}

module.exports.localMatchingService = localMatchingService
module.exports.accountCreationService = accountCreationService

if (process.argv[2] === 'server') {
  const server = http.createServer((request, response) => {
    console.log('Request for: ' + request.method + ' ' + request.url)

    var body = []
    request.on('data', function (chunk) {
      body.push(chunk)
    }).on('end', function () {
      body = Buffer.concat(body).toString()

      if (request.method === 'POST') {
        if (request.headers['content-type'] === 'application/json') {
          var requestJSON = JSON.parse(body)

          switch (request.url) {
            case '/nodejs/matching-service':
              response.statusCode = 200
              response.setHeader('Content-Type', 'application/json')
              response.write(JSON.stringify(localMatchingService(requestJSON)))
              response.end()
              break
            case '/nodejs/account-creation':
              response.statusCode = 200
              response.setHeader('Content-Type', 'application/json')
              response.write(JSON.stringify(accountCreationService(requestJSON)))
              response.end()
              break
            default:
              break
          }
        }
      }
      response.statusCode = 404
      response.end()
    })
  })

  server.listen(port, hostname, () => {
    console.log('Starting local matching service')
  })
}
